### DISCUSSION TITLE
How to use ​/mciRecommendVm

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#1234] [Show and tell]
# 개요
MCI 생성을 위한, VM 구성을 최적화하기 위해 VM 단위의 Spec을 추천하는 방법 설명

# 사용 방법
- [/recommendSpec API](https://cloud-barista.github.io/api/?url=https://converter.swagger.io/api/convert?url=https://raw.githubusercontent.com/cloud-barista/cb-tumblebug/main/src/interface/rest/docs/swagger.json#/%5BMC-Infra%5D%20MCI%20Provisioning%20and%20Management/RecommendSpec)를 사용하면, 사용자가 요청한 사양 및 조건에 맞는 Spec 리스트를 획득할 수 있음
- 필터링 및 우선순위가 적용 가능하며, 다양한 파라미터의 조합이 가능하므로 별도의 옵션을 알고 사용할 필요.

---

## 필터링 옵션

- 기본적으로 [type FilterSpecsByRangeRequest](https://github.com/cloud-barista/cb-tumblebug/blob/main/src/core/model/spec.go#L172) 구조체에 포함된 항목을 기준으로 필터링을 위한 산술, 문자열 일치 조건 적용 가능

ex: `"metric": "providerName"`

### 산술 일치 조건 `<=`, `>=`, `==` (ex: `"operator": "<="`, `"operand": "4"`)

- 	VCPU                `json:"vCPU"`
- 	MemoryGiB           `json:"memoryGiB"`
- 	CostPerHour         `json:"costPerHour"`
- 	AcceleratorCount    `json:"acceleratorCount"`
- 	AcceleratorMemoryGB `json:"acceleratorMemoryGB"`

### 문자열 포함 조건 (ex: `"operand": "gcp"`)

- 	ProviderName        `json:"providerName"`
- 	RegionName          `json:"regionName"`
- 	CspSpecName         `json:"cspSpecName"`
- 	AcceleratorModel    `json:"acceleratorModel"`
- 	AcceleratorType     `json:"acceleratorType"`
- 	Description         `json:"description"`
- 	Description         `json:"description"`

---

## 우선순위 옵션 (현재 metric: location, performance, cost, random 지정 가능)
- location
  - coordinateClose : 제시된 위경도에 가장 가까운 스펙 선정
  - coordinateFair : 제시된 위경도 세트에 가장 공평한 스펙 선정
- performance: 다양한 사양을 기준으로 스펙 선정 (WIP: 현재 [cloudspec.svc](https://github.com/cloud-barista/cb-tumblebug/blob/main/assets/cloudspec.csv)에 evaluationScore01 항목으로만 동작함. Prime Number 계산 성능에 해당하지만, 정확하게 측정된 값은 아님)
- cost: 비용을 기준으로 스펙 선정
- random: 우선순위 없이 무작위로 리스팅

---

## 입력 예시
- filter를 통해 vCPU 수 범위를 지정, AcceleratorModel을 gpu로 지정.
- priority를 통해, metric에 따른 우선순위 결정 (metric: location, performance, cost 지정 가능)
- 상세 가이드: https://github.com/cloud-barista/cb-tumblebug/wiki/Dynamic-and-optimal-mci-provisioning-guide

```
{
  "filter": {
    "policy": [
      {
        "condition": [
          {
            "operand": "4",
            "operator": "<="
          }
        ],
        "metric": "vCPU"
      },
      {
        "condition": [
          {
            "operand": "gpu"
          }
        ],
        "metric": "AcceleratorModel"
      }
    ]
  },
  "limit": "5",
  "priority": {
    "policy": [
      {
        "metric": "location",
        "parameter": [
          {
            "key": "coordinateFair",
            "val": [
              "39.1848/128.2488", "34.0599/131.5093", "36.1672/139.7769"
            ]
          }
        ]
      }
    ]
  }
}
```

**Comment by yunkon-kim** [comment]
@seokho-son 문의드립니다.

아래 Request body를 입력하면 `[]`이 리턴되고 있는데요. 아마도, Priority 부분에서 performance 설정상에 이슈일 것으로 생각됩니다. 관련하여 참고할만한 자료가 있을까요?
```json
{
  "filter": {
    "policy": [
      {
        "condition": [
          {
            "operand": "7",
            "operator": "<="
          },
          {
            "operand": "9",
            "operator": ">="
          }
        ],
        "metric": "vCPU"
      },
      {
        "condition": [
          {
            "operand": "10",
            "operator": "<="
          },
          {
            "operand": "20",
            "operator": ">="
          }
        ],
        "metric": "memoryGiB"
      },
      {
        "condition": [
          {
            "operand": "aws"
          }
        ],
        "metric": "providerName"
      },
      {
        "condition": [
          {
            "operand": "ap-northeast-2"
          }
        ],
        "metric": "regionName"
      }
    ]
  },
  "limit": "5",
  "priority": {
    "policy": [
      {
        "metric": "performance"
      }
    ]
  }
}
```

**Reply by seokho-son** [reply]
@yunkon-kim 일단, 그나마 reference 할만한 문서는 https://github.com/cloud-barista/cb-tumblebug/wiki/Dynamic-and-optimal-mcis-provisioning-guide 입니다. VM 성능 기반 선정의 경우, 평가 데이터를 업데이트하지 않아서, 활용도가 낮은 상태이며, 추가 상세 설정이 현재는 어렵습니다. ^^

다만, 현재 직면하신 문제는 "metric": "performance" 에 의한 이슈가 아니라,
필터 조건식에 방향이 반대로 되어서 인 것 같습니다. (저도 가끔 헷갈립니다 ㅎㅎ)
`[]`는 오류가 아니라, 실제 조건에 해당하는 spec이 하나도 없다는 의미입니다.

- 의미: vCPU가 7보다 작고 9보다 큰 spec (은 없겠죠?)
```
          {
            "operand": "7",
            "operator": "<="
          },
          {
            "operand": "9",
            "operator": ">="
          }
```


아래와 같이 반대로 해보시면, 리스트가 나옵니다.

```
{
  "filter": {
    "policy": [
      {
        "condition": [
          {
            "operand": "7",
            "operator": ">="
          },
          {
            "operand": "9",
            "operator": "<="
          }
        ],
        "metric": "vCPU"
      },
      {
        "condition": [
          {
            "operand": "10",
            "operator": ">="
          },
          {
            "operand": "20",
            "operator": "<="
          }
        ],
        "metric": "memoryGiB"
      },
      {
        "condition": [
          {
            "operand": "aws"
          }
        ],
        "metric": "providerName"
      },
      {
        "condition": [
          {
            "operand": "ap-northeast-2"
          }
        ],
        "metric": "regionName"
      }
    ]
  },
  "limit": "5",
  "priority": {
    "policy": [
      {
        "metric": "performance"
      }
    ]
  }
}
```

**Reply by yunkon-kim** [reply]
@seokho-son  

앗..방향이 반대였네요 ㅠㅠ 

혹시, priority 자체를 없애거나, priority의 policy를 []와 같이 비우는 쪽으로 진행해보겠습니다.

**Reply by seokho-son** [reply]
@yunkon-kim 넵 priority 자체를 없애시면, cost priority 가 기본으로 활성화 됩니다. :)

**Reply by yunkon-kim** [reply]
@seokho-son 무시할 수는 없나보네요 ㅎㅎ;; 그렇다면 performance를 활용하는 쪽으로 가닥을 잡아보겠습니다.

**Reply by seokho-son** [reply]
@yunkon-kim 비밀의 키워드: `random` 을 사용하시면, 의도적으로 셔플링합니다.. ^^;;

**Reply by yunkon-kim** [reply]
@seokho-son 유용할 것 같습니다 :+1: 같이 살펴볼께요.
